rust/rustc.bzl

```diff
@@ -32,20 +32,159 @@ def _get_comp_mode_codegen_opts(ctx, toolchain):
+    transitive_staticlibs = depset()
+    for dep in deps:
+        if hasattr(dep, "crate_info"):
+            # This dependency is a rust_library
+            rlib = dep.crate_info.output
+
+            transitive_rlibs += [rlib]
+            transitive_rlibs += dep.transitive_rlibs
+            transitive_dylibs += dep.transitive_dylibs
+            transitive_staticlibs += dep.transitive_staticlibs
+
+            link_flags += ["--extern " + dep.label.name + "=" + deps_dir + "/" + rlib.basename]
+        elif hasattr(dep, "cc"):
+            # This dependency is a cc_library
+            if not allow_cc_deps:
+                fail("Only rust_library, rust_binary, and rust_test targets can " +
+                     "depend on cc_library")
+
+            transitive_dylibs += dylib_filetype.filter(dep.cc.libs)
+            transitive_staticlibs += staticlib_filetype.filter(dep.cc.libs)
+
+        else:
+            fail("rust_library, rust_binary and rust_test targets can only depend " +
+                 "on rust_library or cc_library targets.")
+
+    link_flags += ["-l static=" + _get_lib_name(lib) for lib in transitive_staticlibs.to_list()]
+    link_flags += ["-l dylib=" + _get_lib_name(lib) for lib in transitive_dylibs.to_list()]
+
+    search_flags = []
+    if transitive_rlibs:
+        search_flags += ["-L dependency={}".format(deps_dir)]
+    if transitive_dylibs or transitive_staticlibs:
+        search_flags += ["-L native={}".format(deps_dir)]
+
+    # Create symlinks pointing to each transitive lib in deps_dir.
+    transitive_libs = transitive_rlibs + transitive_staticlibs + transitive_dylibs
+    for transitive_lib in transitive_libs:
+        setup_cmd += [_create_setup_cmd(transitive_lib, deps_dir, in_runfiles)]
+
+    return struct(
+        link_flags = link_flags,
+        search_flags = search_flags,
+        setup_cmd = setup_cmd,
+        transitive_dylibs = transitive_dylibs,
+        transitive_libs = list(transitive_libs),
+        transitive_rlibs = transitive_rlibs,
+        transitive_staticlibs = transitive_staticlibs,
+    )
+
+_setup_deps = setup_deps
+
+CrateInfo = provider(
+    fields = [
+        "name",
+        "type",
+        "root",
+        "srcs",
+        "deps",
+        "output",
+    ],
+)
+
 # Utility methods that use the toolchain provider.
-def build_rustc_command(
+def rustc_compile_action(
         ctx,
         toolchain,
-        crate_name,
-        crate_type,
-        src,
-        output_dir,
-        depinfo,
+        crate_info,
         output_hash = None,
         rust_flags = []):
     """
     Constructs the rustc command used to build the current target.
     """
+    output_dir = crate_info.output.dirname
+
+    depinfo = _setup_deps(
+        crate_info.deps,
+        crate_info.name,
+        output_dir,
+        toolchain,
+        allow_cc_deps = True,
+    )
+
+    compile_inputs = (
+        crate_info.srcs +
+        ctx.files.data +
```

Comment:
it feels a little weird for this to be in both compile_inputs and runfiles, but oh well

Reply:
Does the default 'data' attr imply runtime data? If so we can add compile_data in short order to differentiate, without dealing w/ embedding the data more directly